src/wallet/wallet_tx_builder.cpp

```diff
@@ -0,0 +1,691 @@
+                        ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                return sendTo.value();
+            },
+            [&](const AccountZTXOPattern& acct) -> ChangeAddress {
+                auto sendTo = pwalletMain->GenerateChangeAddressForAccount(
+                        acct.GetAccountId(),
+                        getAllowedChangePools(acct.GetReceiverTypes()));
+                assert(sendTo.has_value());
+                resolvedPayments.AddPayment(
+                        ResolvedPayment(std::nullopt, sendTo.value(), changeAmount, std::nullopt, true));
+                return sendTo.value();
+            }
+        }, selector.GetPattern());
+    }
+
+    auto ovks = SelectOVKs(selector, spendable);
+
+    return TransactionEffects(
+            anchorConfirmations,
+            spendable,
+            resolvedPayments,
+            changeAddr,
+            fee,
+            ovks.first,
+            ovks.second,
+            anchorHeight);
+}
+
+Payments InputSelection::GetPayments() const {
+    return this->payments;
+}
+
+CAmount WalletTxBuilder::DefaultDustThreshold() const {
+    CKey secret{CKey::TestOnlyRandomKey(true)};
+    CScript scriptPubKey = GetScriptForDestination(secret.GetPubKey().GetID());
+    CTxOut txout(CAmount(1), scriptPubKey);
+    return txout.GetDustThreshold(minRelayFee);
+}
+
+SpendableInputs WalletTxBuilder::FindAllSpendableInputs(
+        const ZTXOSelector& selector,
+        int32_t minDepth) const
+{
+    LOCK2(cs_main, wallet.cs_wallet);
+    return wallet.FindSpendableInputs(selector, minDepth, std::nullopt);
+}
+
+InputSelectionResult WalletTxBuilder::ResolveInputsAndPayments(
+        const ZTXOSelector& selector,
+        SpendableInputs& spendableMut,
+        const std::vector<Payment>& payments,
+        const CChain& chain,
+        TransactionStrategy strategy,
+        CAmount fee,
+        int anchorHeight) const
+{
+    LOCK2(cs_main, wallet.cs_wallet);
+
+    // Determine the target totals
+    CAmount sendAmount{0};
+    for (const auto& payment : payments) {
+        sendAmount += payment.GetAmount();
+    }
+    CAmount targetAmount = sendAmount + fee;
+
+    // This is a simple greedy algorithm to attempt to preserve requested
+    // transactional privacy while moving as much value to the most recent pool
+    // as possible.  This will also perform opportunistic shielding if the
+    // transaction strategy permits.
+
+    CAmount maxSaplingAvailable = spendableMut.GetSaplingTotal();
+    CAmount maxOrchardAvailable = spendableMut.GetOrchardTotal();
+    uint32_t orchardOutputs{0};
+
+    // we can only select Orchard addresses if there are sufficient non-Sprout
+    // funds to cover the total payments + fee.
+    bool canResolveOrchard =
+        params.GetConsensus().NetworkUpgradeActive(anchorHeight, Consensus::UPGRADE_NU5)
+        && spendableMut.Total() - spendableMut.GetSproutTotal() >= targetAmount;
```

Comment:
Non-blocking: this means that `canResolveOrchard = false` if the wallet has no Sprout funds, but also has insufficient funds for the transaction as a whole. This may result in somewhat confusing errors being reported?